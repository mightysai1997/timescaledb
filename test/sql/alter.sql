-- DROP a table's column before making it a hypertable
CREATE TABLE alter_before(id serial, time timestamp, temp float, colorid integer);
ALTER TABLE alter_before DROP COLUMN id;

SELECT create_hypertable('alter_before', 'time', chunk_time_interval => 2628000000000);

INSERT INTO alter_before VALUES ('2017-03-22T09:18:22', 23.5, 1);

SELECT * FROM alter_before;

-- Show that deleted column is marked as dropped and that attnums are
-- now different for the root table and the chunk
SELECT c.relname, a.attname, a.attnum FROM pg_attribute a, pg_class c
WHERE a.attrelid = c.oid
AND (c.relname LIKE '_hyper_1%_chunk' OR c.relname = 'alter_before')
AND a.attnum > 0
ORDER BY c.relname, a.attnum;

-- DROP a table's column after making it a hypertable and having data
CREATE TABLE alter_after(id serial, time timestamp, temp float, colorid integer);
SELECT create_hypertable('alter_after', 'time', chunk_time_interval => 2628000000000);

-- Create first chunk
INSERT INTO alter_after (time, temp, colorid) VALUES ('2017-03-22T09:18:22', 23.5, 1);

ALTER TABLE alter_after DROP COLUMN id;

-- Creating new chunks after dropping a column should work just fine
INSERT INTO alter_after VALUES ('2017-03-22T09:18:23', 21.5, 1),
                               ('2017-05-22T09:18:22', 36.2, 2),
                               ('2017-05-22T09:18:23', 15.2, 2);

-- Make sure tuple conversion also works with COPY
\COPY alter_after FROM 'data/alter.tsv' NULL AS '';

-- Data should look OK
SELECT * FROM alter_after;

-- Show that attnums are different for chunks created after DROP
-- column
SELECT c.relname, a.attname, a.attnum FROM pg_attribute a, pg_class c
WHERE a.attrelid = c.oid
AND (c.relname LIKE '_hyper_2%_chunk' OR c.relname = 'alter_after')
AND a.attnum > 0
ORDER BY c.relname, a.attnum;

-- Add an ID column again
ALTER TABLE alter_after ADD COLUMN id serial;

INSERT INTO alter_after (time, temp, colorid) VALUES ('2017-08-22T09:19:14', 12.5, 3);

SELECT c.relname, a.attname, a.attnum FROM pg_attribute a, pg_class c
WHERE a.attrelid = c.oid
AND (c.relname LIKE '_hyper_2%_chunk' OR c.relname = 'alter_after')
AND a.attnum > 0
ORDER BY c.relname, a.attnum;

SELECT * FROM alter_after;
