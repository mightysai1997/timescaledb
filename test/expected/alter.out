-- DROP a table's column before making it a hypertable
CREATE TABLE alter_before(id serial, time timestamp, temp float, colorid integer);
ALTER TABLE alter_before DROP COLUMN id;
SELECT create_hypertable('alter_before', 'time', chunk_time_interval => 2628000000000);
NOTICE:  adding NOT NULL constraint to column "time"
 create_hypertable 
-------------------
 
(1 row)

INSERT INTO alter_before VALUES ('2017-03-22T09:18:22', 23.5, 1);
SELECT * FROM alter_before;
           time           | temp | colorid 
--------------------------+------+---------
 Wed Mar 22 09:18:22 2017 | 23.5 |       1
(1 row)

-- Show that deleted column is marked as dropped and that attnums are
-- now different for the root table and the chunk
SELECT c.relname, a.attname, a.attnum FROM pg_attribute a, pg_class c
WHERE a.attrelid = c.oid
AND (c.relname LIKE '_hyper_1%_chunk' OR c.relname = 'alter_before')
AND a.attnum > 0;
     relname      |           attname            | attnum 
------------------+------------------------------+--------
 alter_before     | ........pg.dropped.1........ |      1
 alter_before     | time                         |      2
 alter_before     | temp                         |      3
 alter_before     | colorid                      |      4
 _hyper_1_1_chunk | time                         |      1
 _hyper_1_1_chunk | temp                         |      2
 _hyper_1_1_chunk | colorid                      |      3
(7 rows)

-- DROP a table's column after making it a hypertable and having data
CREATE TABLE alter_after(id serial, time timestamp, temp float, colorid integer);
SELECT create_hypertable('alter_after', 'time', chunk_time_interval => 2628000000000);
NOTICE:  adding NOT NULL constraint to column "time"
 create_hypertable 
-------------------
 
(1 row)

-- Create first chunk
INSERT INTO alter_after (time, temp, colorid) VALUES ('2017-03-22T09:18:22', 23.5, 1);
ALTER TABLE alter_after DROP COLUMN id;
-- Creating new chunks after dropping a column should work just fine
INSERT INTO alter_after VALUES ('2017-03-22T09:18:23', 21.5, 1),
                               ('2017-05-22T09:18:22', 36.2, 2),
                               ('2017-05-22T09:18:23', 15.2, 2);
-- Make sure tuple conversion also works with COPY
\COPY alter_after FROM 'data/alter.tsv' NULL AS '';
-- Data should look OK
SELECT * FROM alter_after;
           time           | temp | colorid 
--------------------------+------+---------
 Wed Mar 22 09:18:22 2017 | 23.5 |       1
 Wed Mar 22 09:18:23 2017 | 21.5 |       1
 Mon May 22 09:18:22 2017 | 36.2 |       2
 Mon May 22 09:18:23 2017 | 15.2 |       2
 Tue Aug 22 09:19:22 2017 | 21.4 |       3
 Wed Aug 23 09:20:17 2017 | 31.5 |       2
(6 rows)

-- Show that attnums are different for chunks created after DROP
-- column
SELECT c.relname, a.attname, a.attnum FROM pg_attribute a, pg_class c
WHERE a.attrelid = c.oid
AND (c.relname LIKE '_hyper_2%_chunk' OR c.relname = 'alter_after')
AND a.attnum > 0;
     relname      |           attname            | attnum 
------------------+------------------------------+--------
 alter_after      | ........pg.dropped.1........ |      1
 alter_after      | time                         |      2
 alter_after      | temp                         |      3
 alter_after      | colorid                      |      4
 _hyper_2_2_chunk | ........pg.dropped.1........ |      1
 _hyper_2_2_chunk | time                         |      2
 _hyper_2_2_chunk | temp                         |      3
 _hyper_2_2_chunk | colorid                      |      4
 _hyper_2_3_chunk | time                         |      1
 _hyper_2_3_chunk | temp                         |      2
 _hyper_2_3_chunk | colorid                      |      3
 _hyper_2_4_chunk | time                         |      1
 _hyper_2_4_chunk | temp                         |      2
 _hyper_2_4_chunk | colorid                      |      3
(14 rows)

-- Add an ID column again
ALTER TABLE alter_after ADD COLUMN id serial;
INSERT INTO alter_after (time, temp, colorid) VALUES ('2017-08-22T09:19:14', 12.5, 3);
SELECT c.relname, a.attname, a.attnum FROM pg_attribute a, pg_class c
WHERE a.attrelid = c.oid
AND (c.relname LIKE '_hyper_2%_chunk' OR c.relname = 'alter_after')
AND a.attnum > 0;
     relname      |           attname            | attnum 
------------------+------------------------------+--------
 alter_after      | ........pg.dropped.1........ |      1
 alter_after      | time                         |      2
 alter_after      | temp                         |      3
 alter_after      | colorid                      |      4
 alter_after      | id                           |      5
 _hyper_2_2_chunk | ........pg.dropped.1........ |      1
 _hyper_2_2_chunk | time                         |      2
 _hyper_2_2_chunk | temp                         |      3
 _hyper_2_2_chunk | colorid                      |      4
 _hyper_2_2_chunk | id                           |      5
 _hyper_2_3_chunk | time                         |      1
 _hyper_2_3_chunk | temp                         |      2
 _hyper_2_3_chunk | colorid                      |      3
 _hyper_2_3_chunk | id                           |      4
 _hyper_2_4_chunk | time                         |      1
 _hyper_2_4_chunk | temp                         |      2
 _hyper_2_4_chunk | colorid                      |      3
 _hyper_2_4_chunk | id                           |      4
(18 rows)

SELECT * FROM alter_after;
           time           | temp | colorid | id 
--------------------------+------+---------+----
 Wed Mar 22 09:18:22 2017 | 23.5 |       1 |  1
 Wed Mar 22 09:18:23 2017 | 21.5 |       1 |  2
 Mon May 22 09:18:22 2017 | 36.2 |       2 |  3
 Mon May 22 09:18:23 2017 | 15.2 |       2 |  4
 Tue Aug 22 09:19:22 2017 | 21.4 |       3 |  5
 Wed Aug 23 09:20:17 2017 | 31.5 |       2 |  6
 Tue Aug 22 09:19:14 2017 | 12.5 |       3 |  7
(7 rows)

